k2pow-service/src/job_manager.rs

```diff
@@ -0,0 +1,213 @@
+use crate::{create_thread_pool, PoW};
+use post::pow::Prover;
+use std::collections::HashMap;
+use std::sync::Mutex;
+use thiserror::Error;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum JobError {
+    #[error("too many active jobs")]
+    TooManyJobs,
+}
+
+/// JobStatus is used to expose job state to external callers
+#[derive(Clone, Debug, PartialEq)]
+pub enum JobStatus {
+    Created,
+    InProgress,
+    Done(Result<u64, String>),
+}
+
+#[derive(Debug)]
+enum JobState {
+    InProgress(std::thread::JoinHandle<Result<u64, post::pow::Error>>),
+    Done(Result<u64, String>),
+}
+
+#[derive(Hash, Eq, PartialEq, Debug, Clone)]
+pub struct Job {
+    pub nonce_group: u8,
+    pub challenge: [u8; 8],
+    pub difficulty: [u8; 32],
+    pub miner: [u8; 32],
+}
+
+#[cfg_attr(test, mockall::automock)]
+pub trait GetOrCreate {
+    fn get_or_create(&self, job: Job) -> Result<JobStatus, JobError>;
+}
+pub struct JobManager {
+    jobs: Mutex<HashMap<Job, JobState>>,
+    cores: u8,
+    randomx_mode: crate::RandomXMode,
+    randomx_large_pages: bool,
+}
+
+impl JobManager {
+    pub fn new(cores: u8, randomx_mode: crate::RandomXMode, randomx_large_pages: bool) -> Self {
+        JobManager {
+            jobs: Mutex::new(HashMap::new()),
+            cores,
+            randomx_mode,
+            randomx_large_pages,
+        }
+    }
+    fn check_finished(&self, job: &Job) {
+        // there's probably a better way to write this with interior mutability on the
+        // value that the HashMap stores (like with a RefCell). For now, because the
+        // joining the handle consumes the value, borrowing doesn't work here and we need
+        // to remove the value, use it, then put it back into the map.
+        let mut hs = self.jobs.lock().unwrap();
+        let entry = hs.remove_entry(job);
```

Comment:
AFAIR, I used an `Option<JoinHandle>` for this reason before so I could `take()` handle out of it and later put it back.